localstack/services/sqs/provider.py

```diff
@@ -0,0 +1,1243 @@
+        original_message = None
+        original_message_group = self.deduplication.get(message_group_id)
+        if original_message_group:
+            original_message = original_message_group.get(dedup_id)
+        if (
+            original_message
+            and not original_message.deleted
+            and original_message.priority + DEDUPLICATION_INTERVALL_IN_SEC > qm.priority
+        ):
+            message["MessageId"] = original_message.message["MessageId"]
+        else:
+            self.visible.put_nowait(qm)
+            if not original_message_group:
+                self.deduplication[message_group_id] = dict()
+            self.deduplication[message_group_id][message_deduplication_id] = qm
+
+    def _assert_queue_name(self, name):
+        if not name.endswith(".fifo"):
+            raise InvalidParameterValue(
+                "Can only include alphanumeric characters, hyphens, or underscores. 1 to 80 in length"
+            )
+        # The .fifo suffix counts towards the 80-character queue name quota.
+        queue_name = name[:-5] + "_fifo"
+        super()._assert_queue_name(queue_name)
+
+    def validate_queue_attributes(self, attributes):
+        valid = [k[1] for k in inspect.getmembers(QueueAttributeName)]
+
+        for k in attributes.keys():
+            if k not in valid:
+                raise InvalidAttributeName("Unknown attribute name %s" % k)
+        # Special Cases
+        if attributes.get(QueueAttributeName.FifoQueue).lower() != "true":
+            raise InvalidAttributeValue(
+                "Invalid value for the parameter FifoQueue. Reason: Modifying queue type is not supported."
+            )
+
+
+class InflightUpdateWorker:
+    """
+    Regularly re-queues inflight messages whose visibility timeout has expired.
+
+    FIXME: very crude implementation. it would be better to have event-driven communication.
+    """
+
+    queues: Dict[QueueKey, SqsQueue]
+
+    def __init__(self, queues: Dict[QueueKey, SqsQueue]) -> None:
+        super().__init__()
+        self.queues = queues
+        self.running = False
+        self.thread: Optional[FuncThread] = None
+
+    def start(self):
+        if self.thread:
+            return
+
+        def _run(*_args):
+            self.running = True
+            self.run()
+
+        self.thread = start_thread(_run)
+
+    def stop(self):
+        if self.thread:
+            self.thread.stop()
+
+        self.running = False
+        self.thread = None
+
+    def run(self):
+        while self.running:
+            time.sleep(1)
+            for queue in self.queues.values():
+                queue.requeue_inflight_messages()
+
+
+class SqsProvider(SqsApi, ServiceLifecycleHook):
+    """
+    LocalStack SQS Provider.
+
+    LIMITATIONS:
+        - Calculation of message attribute MD5 hashes
+        - Pagination of results (NextToken)
+        - Sequence numbering
+        - Delivery guarantees
+        - FIFO/Standard queue semantics
```

Comment:
seems a bit too ambiguous. As far as I can see the FIFO/Standard queue separation is implemented now.